share/vss/vss.go

```diff
@@ -0,0 +1,579 @@
+// dealer is malicious. In that case, there needs to be some way for Verifiers to
+// check the correctness of the shares they receive. This is the scheme goal.
+// 1) The dealer send a Deal to every verifiers using `Deals()`
+// 2) Each verifier process the Deal with `ProcessDeal`. They can either return:
+//    - an Approval, which means the deal seems correct
+//    - an Complaint, which means the deal is not correct and the dealer might
+//    be malicious.
+// 3) The dealer can respond to each Complaint by a Justification, basically
+// revealing the share he sent out to the original "complainer"
+// 4) All verifiers accept the deal iif there has been at least t Approval. They
+// refuse the deal if there has been more than t-1 complaints OR if a
+// Justification is wrong.
+// At the end of the scheme, all verifiers can re-unite their Share to
+// reconstruct the original secret. Only t out of n verifiers are needed.
+package vss
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/share"
+	"github.com/dedis/crypto/sign"
+)
+
+// Dealer is responsible for creating the shares, distribute them and reply to
+// any Complaints.
+type Dealer struct {
+	suite abstract.Suite
+
+	reader cipher.Stream
+	long   abstract.Scalar
+
+	secret abstract.Scalar
+
+	verifiers []abstract.Point
+	// threshold of shares that is needed to reconstruct the secret
+	t int
+	// sessionID is a unique identifier for the whole session of the scheme
+	sessionID []byte
+
+	// list of deals this Dealer has generated
+	deals []*Deal
+	*aggregator
+}
+
+// Deal is sent by the Dealer to each participants.
+type Deal struct {
+	SessionID   []byte
+	SecShare    *share.PriShare
+	RndShare    *share.PriShare
+	T           uint32
+	Commitments []abstract.Point
+	Signature   []byte
+}
+
+// Complaint is a message that must be broadcasted to every verifiers when
+// a verifier receives an invalid Deal.
+type Complaint struct {
+	SessionID []byte
+	Index     uint32
+	Deal      *Deal
+	Signature []byte
+}
+
+// Justification is a message that is broadcasted by the Dealer in response to
+// a Complaint. It contains the original Complaint as well as the shares
+// distributed to the complainer.
+type Justification struct {
+	Complaint *Complaint
+	Deal      *Deal
+}
+
+// Approval is a message that is sent if a verifier approves the Deal he
+// received from the Dealer.
+type Approval struct {
+	SessionID []byte
+	Index     uint32
+	Signature []byte
+}
+
+// NewDealer returns a Dealer capable of leading the secret sharing scheme. It
+// does not have to be trusted by other Verifiers. The security parameter t is
+// the number of shares required to reconstruct the secret. It must be between
+// (len(verifiers) + 1)/2 <= t <= len(verifiers).
+func NewDealer(suite abstract.Suite, longterm, secret abstract.Scalar, verifiers []abstract.Point, r cipher.Stream, t int) (*Dealer, error) {
```

Comment:
> NewDealer creates and returns a dealer used to lead a VSS run among n verifiers and a secret sharing threshold t.

As discussed earlier today, I am not sure if we should enforce that `t` should be in the above range.